Final 1st draft for Specifying Representation, modified content for Querying Representation (#549)
AdaCore/learn: content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

Note that :ada:`Address` is the type of the result of the query attribute
:ada:`Address`.

We mentioned potentially needing to swap bytes in networking 
communications software, due to the differences in the "endianness" of 
the machines communicating. That characteristic can be determined via a 
constant declared in package :ada:`System` as follows: 

.. code-block:: ada

   type Bit_Order is (High_Order_First, Low_Order_First);
   Default_Bit_Order : constant Bit_Order := implementation-defined;

:ada:`High_Order_First` corresponds to "Big Endian" and 
:ada:`Low_Order_First` to "Little Endian." On a Big Endian machine, bit 
0 is the most significant bit. On a Little Endian machine, bit 0 is the 
least significant bit. 

> Strictly speaking, this constant gives us the default order for bits 

Reviewer: We've actually removed references to big/little endianess from the previous book (see https://github.com/AdaCore/learn/pull/447/files#r532750100 on line 919.)

Reviewer: I assume it's ok here though

Author: I left them off of the discussion on record rep clauses, as being beyond the scope of the course. Butt this is just how to determine the endina-ness, so IMO we should have it.

within storage elements in record representation clauses, not the order 
of bytes within words. However, we can usually use it for the byte order 

> too. In particular, if Word_Size is greater than Storage_Unit, a word 

Reviewer: ```suggestion
too. In particular, if :ada:`Word_Size` is greater than :ada:`Storage_Unit`, a word 
```

necessarily consists of multiple storage elements, so the default bit 
ordering is the same as the ordering of storage elements in a word. 

Let's take that example of swapping the bytes in a received Ethernet 
packet. The "wire" format is Big Endian so if we are running on a Little 
Endian machine we must swap the bytes received.

> Suppose we want retrieve typed values from a 

Reviewer: ```suggestion
Suppose we want to retrieve typed values from a 
```

> given buffer or bytes. We get the bytes from the buffer into a variable named Value, of the type of interest, and then swap those bytes within Value if necessary.

Reviewer: ```suggestion
given buffer or bytes. We get the bytes from the buffer into a variable named :ada:`Value`, of the type of interest, and then swap those bytes within :ada:`Value` if necessary.
```

.. code-block:: ada

      ...
   begin
      Value := ...
      
      if Default_Bit_Order /= High_Order_First then -- we're not on a Big Endian machine
         Value := Byte_Swapped (Value);
      end if;
   end Retrieve_4_Bytes;

> We have elided the code that gets the bytes into Value, for the sake of 
> simplicity. How the bytes are actually swapped by function Byte_Swapped 

Reviewer: ```suggestion
We have elided the code that gets the bytes into :ada:`Value`, for the sake of 
simplicity. How the bytes are actually swapped by function :ada:`Byte_Swapped`
```

Querying Representation Choices
-------------------------------

As we mentioned in the introduction, in low-level programming the 
hardware and the compiler's representation choices can come to the 
forefront. You can, therefore, query many such choices. 

For example, let's say we want to query the addresses of some objects

> because we are calling the imported C "memcpy" function. That function

Reviewer: ```suggestion
because we are calling the imported C c:`memcpy` function. That function
```

             Source /= Destination       and then
             not Overlapping (Destination, Source, Length),
     Post => MemCopy'Result = Destination;
   --  Copies Length bytes from the object designated by Source to the object
   --  designated by Destination. 

The three aspects that do the importing are specified after the reserved 
word :ada:`with` but can be ignored for this discussion. We'll talk 
about them later. The preconditions make explicit the otherwise implicit 
requirements for the arguments passed to memcpy, and the postcondition 
specifies the expected result returned from a successful call. Neither 
the preconditions nor the postconditions are required for importing 
external entities but they are good "guard-rails" for using those 
entities. If we call it incorrectly the precondition will inform us, and 
likewise, if we misunderstand the result the postcondition will let us 
know (at least to the extent that the return value does that). 

For a sample call to our imported routine, imagine that we have a 
procedure that copies the bytes of a String parameter into a Buffer 
parameter, which is just a contiguous array of bytes. We need to tell 

> MemCopy the addresses of the arguments passed so we apply the attribute 

Reviewer: ```suggestion
:ada:`MemCopy` the addresses of the arguments passed so we apply the attribute 
```

accordingly: 

.. code-block:: ada

   procedure Put (This : in out Buffer; Start : Index; Value : String) is
      Result : System.Address with Unreferenced;
   begin
      Result := MemCopy (Destination => This (Start)'Address,
                         Source      => Value'Address,
                         Length      => Value'Length);
   end Put;
   
The order of the address parameters is easily confused so we use the 
named association format for specifying the actual parameters in the 
call. 

> Although we assign Result we don't otherwise use it, so we tell the

Reviewer: ```suggestion
Although we assign :ada:`Result` we don't otherwise use it, so we tell the
```

compiler this is not a mistake via the :ada:`Unreferenced` aspect. And if we
do turn around and reference it the compiler will complain, as it should.

> (We don't show the preconditions for Put, but they would have specified 
> that Start must be a valid index into This particular Buffer, and that 
> there must be room in the Buffer argument for the number of bytes in 
> Value when starting at the Start index, so that we don't copy past the 
> end of the Buffer argument.) 

Reviewer: ```suggestion
(We don't show the preconditions for :ada:`Put`, but they would have specified 
that :ada:`Start` must be a valid index into this particular buffer, and that 
there must be room in the :ada:`Buffer` argument for the number of bytes in 
:ada:`Value` when starting at the :ada:`Start` index, so that we don't copy past the 
end of the :ada:`Buffer` argument.) 
```

that any storage element's address will suffice. If the machine has no 
particular natural alignments, then all type Alignments will probably 
be 1 by default. That would be somewhat rare today, though, because 
modern processors usually have comparatively strict alignment requirements. 

We can ask for the amount of storage associated with various entities. 
For example, when applied to a task, :ada:`'Storage_Size` tells us the 
number of storage elements reserved for the task's execution. The value 
includes the size of the task's stack, if it has one. We aren't told if 
other required storage, used internally in the implementation, is also 
included in this number. Often that other storage is not included in 
this number, but it could be. 

:ada:`Storage_Size` is also defined for access types. The meaning is a 
little complicated. Access types can be classified into those that 
designate only variables ("access-to-object") and those that can also 
designate constants. Constants are never allocated dynamically so we can 
ignore them. Each access-to-object type has an associated storage pool. 
The storage allocated by :ada:`new` comes from the pool, and instances 

> of Unchecked_Deallocation return storage to the pool. 

Reviewer: ```suggestion
of :ada:`Unchecked_Deallocation` return storage to the pool. 
```

call. 

Although we assign :ada:`Result` we don't otherwise use it, so we tell the
compiler this is not a mistake via the :ada:`Unreferenced` aspect. And if we
do turn around and reference it the compiler will complain, as it should.

(We don't show the preconditions for :ada:`Put`, but they would have specified 
that :ada:`Start` must be a valid index into this particular buffer, and that 
there must be room in the :ada:`Buffer` argument for the number of bytes in 
:ada:`Value` when starting at the :ada:`Start` index, so that we don't copy past the 
end of the :ada:`Buffer` argument.) 

There are other characteristics we might want to query too. 

We might want to ask the compiler what alignment it chose for a given 
object (or type, for all such objects). For a type, an :ada:`Alignment` of zero 
means that objects of the type are not normally aligned on a storage 
element boundary at all. That could happen if the type is packed down 
into a composite object, such as an array of :ada:`Booleans`. We'll discuss 
"packing" soon. More commonly, the smallest likely value is 1, meaning 
that any storage element's address will suffice. If the machine has no 

> particular natural alignments, then all type :ada:`Alignment`s will probably 

Reviewer: ```suggestion
particular natural alignments, then all type alignments will probably 
```

If we apply :ada:`'Size` to a type, the resulting value depends on the
kind of type.

For scalar types, the attribute returns the *minimum* number of bits 
required to represent all the values of the type. Here's a diagram 
showing what the category "scalar types" includes: 

> .. image:: images/scalar_types_tree.png.png

Reviewer: ```suggestion
.. image:: images/scalar_types_tree.png
```

*parameter* holding the value to be sent. That will tell you how many
bits are really needed.

> One last point. GNAT, and now Ada 202x, define an attribute named 

Reviewer: ```suggestion
One last point: GNAT, and now Ada 202x, define an attribute named 
```

:ada:`Object_Size`. It does just what the name suggests: what 
:ada:`'Size` does when applied to objects rather than types. GNAT also 
defines another attribute, named :ada:`Value_Size`, that does what 
:ada:`'Size` does when applied to types. The former is far more useful 
so Ada has standardized it. 

Specifying Representation
-------------------------

Recall that we said :ada:`Boolean'Size` is required to be 1, and that

> stand-alone objects of type Boolean are very likely allocated some

Reviewer: ```suggestion
stand-alone objects of type :ada:`Boolean` are very likely allocated some
```

representation characteristics of types, and thus objects of those 
types, including their bit-wise layouts. It also allows us to specify 
the representation of individual objects. You should understand, though, 
that the compiler is not required to do what you ask. That liberty is 
unavoidable because you might ask the impossible. If you specify that 
the array of 16 Booleans is to be represented completely in 15 bits, 
what can the compiler do? Rejecting that specification is the only 
reasonable response. But if you specify something possible the compiler 
is going to do what you ask, absent some compelling reason to the 
contrary. 

With that in mind, let's examine setting the size for types.

So, how do we specify that the want our array of 16 Boolean components 
to be allocated one bit per component, for a total allocation of 16 
bits? There are a couple of ways, one somewhat better than the other. 

First, you can ask that the compiler "pack" the components into as small
a number of bits as it can:

.. code-block:: ada

   type Bits16 is array (0 .. 15) of Boolean with
     Pack;

That likely does what you want: Bits16'Size will probably be 16.

> But realize that the Pack aspect (and corresponding pragma) is merely a

Reviewer: ```suggestion
But realize that the :ada:`Pack` aspect (and corresponding pragma) is merely a
```

request that the compiler do its best to minimize the number of bits
allocated, not necessarily that it do exactly what you expected or
required.

We could set the size of the entire array type:

.. code-block:: ada

   type Bits16 is array (0 .. 15) of Boolean with
     Size => 16;

> But the language standard says that a Size clause on array and record 

Reviewer: ```suggestion
But the language standard says that a :ada:`Size` clause on array and record 
```

types should not affect the internal layout of their components. That's 
Implementation Advice, so not normative, but implementations are really 
expected to follow the advice, absent some compelling reason. That's 

> what the Pack aspect, record_representation_clauses, and Component_Size 
> clauses are for. (We'll talk about record_representation_clauses 

Reviewer: ```suggestion
what the :ada:`Pack` aspect, :ada:`record_representation_clauses`, and :ada:`Component_Size` 
clauses are for. (We'll talk about :ada:`record_representation_clauses` 
```

momentarily.) That said, at least one other vendor's compiler would have 
changed the size of the array type because of the Size clause, so GNAT 
defines a configuration pragma named Implicit_Packing that overrides the 
default behavior. With that pragma applied, the Size clause would 
compile and suffice to make the overall size be 16. That's a 
vendor-defined pragma though, so not portable. 

Therefore, the best way to set the size for the array type is to set the 

> size of the individual components, via the Component_Size aspect as the 

Reviewer: ```suggestion
size of the individual components, via the :ada:`Component_Size` aspect as the 
```

specifying the address of an object. 

Let's say you want to access the individual bits of what is usually 
treated as a simple signed integer. One way to do that is to express an 
"overlay," in which an object of one type is placed at the same memory 
location as a distinct object of a distinct type, thus overlaying one 
object over the other. Doing so allows you to interact with that memory 
location in more than one way: one way per type, specifically. This is 
known as `type punning <https://en.wikipedia.org/wiki/Type_punning>`_ in 
computer programming. There are other ways to achieve that effect as 
well, but realize that you're circumventing the static strong typing 
used by Ada to protect us from ourselves and from others. Use it with 
care! 

.. code-block:: ada

   type Bits32 is array (0 .. 31) of Boolean with
     Component_Size => 1;

   X : Integer;
   Y : Bits32 with Address => X'Address;

We can query the addresses of objects, and other things too, but 
objects, especially variables, are the most common case. In the above, 
we say :ada:`X'Address` to query the starting address of object 
:ada:`X`. With that information we know what address to specify for our 

> bit-mask overlay Y. Now X and Y are aliases, and therefore we can 

Reviewer: ```suggestion
bit-mask overlay :ada:`Y`. Now :ada:`X` and :ada:`Y` are aliases, and therefore we can 
```

wide, so that we know what size array to use for the bit-mask. The 
answer is that you just have to know the target well when doing 
low-level programming, and that portability is not the controlling 
requirement. The hardware becomes much more visible, as we mentioned. 

That said, you could at least verify the assumption:

.. code-block:: ada

   pragma Compile_Time_Error (Integer'Object_Size /= 32, "Integers expected to be 32 bits");
   X : Integer;
   Y : Bits32 with Address => X'Address;

That's a vendor-defined pragma so this is not fully portable. It isn't 
an unusual pragma, though, so at least you can probably get the same 
functionality even if the pragma name varies. 

Let's return, momentarily, to setting the size of entities, but now let's 
focus on setting the size of objects. 

We've said that the size of an object is not necessarily the same as the 
size of the object's type. The object size won't be smaller, but it 
could be larger. Why? For a stand-alone object or a parameter, most 
implementations will round the size up to a storage element boundary, or 
more, so the object size might be greater than that of the type. Think 

> back to Boolean, where Size is required to be 1, but stand-alone objects 

Reviewer: ```suggestion
back to :ada:`Boolean`, where :ada:`Size` is required to be 1, but stand-alone objects 
```

are probably allocated 8 bits, i.e., an entire storage element (on our 
hypothetical byte-addressed machine). 

Likewise, recall that numeric type declarations are mapped to underlying 
hardware numeric types. These underlying numeric types provide at least 
the capabilities we request with our type declarations, e.g., the range 
or number of digits, perhaps more. But the mapped numeric hardware type 
cannot provide less than requested. If there is no underlying hardware 
type with at least our requested capabilities, our declarations won't compile. 
That mapping means that specifying the size of a numeric type doesn't  
necessarily affect the size of objects of the type. That numeric 
hardware type is the size that it is, and is fixed by the hardware. 

For example, let's say we have this declaration:

.. code-block:: ada

   type Device_Register is range 0 .. 2**5 - 1 with Size => 5;
   
That will compile successfully, because there will be a signed integer 
hardware type with at least that range. (Not necessarily, legally 
speaking, but realistically speaking, there will be such a hardware 
type.) Indeed, it may be an 8-bit signed integer, in which case 

> Device_Register'Size will give us 5, but objects of the type will have a 

Reviewer: ```suggestion
:ada:`Device_Register'Size` will give us 5, but objects of the type will have a 
```

> size of 8, unavoidably, even though we set Size to 5.

Reviewer: ```suggestion
size of 8, unavoidably, even though we set :ada:`Size` to 5.
```

The difference between the type and object sizes can lead to potentially 
problematic code: 

.. code-block:: ada

   type Device_Register is range 0 .. 2**8 - 1 with Size => 8;
   My_Device : Device_Register
     with Address => To_Address (...); 

The code compiles successfully, and tries to map a byte to a hardware 
device that is physically connected to one storage element in the 
processor memory space. The actual address is elided as it is not 
important here. 

That code might work too, but it might not. We might think that 

> My_Device'Size is 8, and that My_Device'Address points at an 8-bit 

Reviewer: ```suggestion
:ada:`My_Device'Size` is 8, and that :ada:`My_Device'Address` points at an 8-bit 
```

location. However, this isn't necessarily so, as we saw with the 
supposedly 5-bit example earlier. Maybe the smallest signed integer the 
hardware has is 16-bits wide. The code would compile because a 16-bit 
signed numeric type can certainly handle the 8-bit range requested. 

> My_Device'Size would be then 16, and because 'Address gives us the 
> *starting* storage element, My_Device'Address might designate the 
> high-order byte of the overall 16-bit object. When the compiler reads 
> the two bytes for My_Device what will happen? One of the bytes will be 

Reviewer: ```suggestion
:ada:`My_Device'Size` would be then 16, and because :ada:`'Address` gives us the 
*starting* storage element, :ada:`My_Device'Address` might designate the 
high-order byte of the overall 16-bit object. When the compiler reads 
the two bytes for :ada:`My_Device` what will happen? One of the bytes will be 
```

the data presented by the hardware device mapped to the memory. The 
other byte will contain undefined junk, whatever happens to be in the 
memory cell at the time. We might have to debug the code a long time to 
identify that as the problem. More likely we'll conclude we have a 
failed device. 

The correct way to write the code is to specify the size of the 
object instead of the type: 

.. code-block:: ada

   type Device_Register is range 0 .. 2**8 - 1; 
   My_Device : Device_Register with
     Size => 8,
     Address => To_Address (...); 

If the compiler cannot support stand-alone 8-bit objects, the code
won't compile. 

> Alternatively, we could change the earlier Size clause on the 
> type to apply Object_Size instead: 

Reviewer: ```suggestion
Alternatively, we could change the earlier :ada:`Size` clause on the 
type to apply :ada:`Object_Size` instead: 
```

.. code-block:: ada

   type Device_Register is range 0 .. 2**8 - 1 with Object_Size => 8; 
   My_Device : Device_Register with
     Address => To_Address (...); 

The choice between the two approaches comes down to personal preference. 
With either approach, if the implementation cannot support 8-bit 
stand-alone objects, we find out that there is a problem at 
compile-time. That's always cheaper than debugging. 

> You might conclude that setting the Size for a type serves no purpose. 

Reviewer: ```suggestion
You might conclude that setting the :ada:`Size` for a type serves no purpose. 
```

That's not an unreasonable conclusion, given what you've seen, but in 
fact there are reasons to do so. However, there are only a few specific 
cases so we will save the reasons for the discussions of the specific 
cases. 

> There is one general case, though, for setting the 'Size of a type. 

Reviewer: ```suggestion
There is one general case, though, for setting the :ada:`'Size` of a type. 
```

Specifically, you may want to specify the size that you think is the 
minimum possible, and you want the compiler to confirm that belief. This 
would be one of the so-called "confirming" representation clauses, in 
which the representation detail is what the compiler would have chosen 
anyway, absent the specification. You're not actually changing anything, 

> you're just getting confirmation via whether or not the compiler accepts 

Reviewer: ```suggestion
you're just getting confirmation via :ada:`Size` whether or not the compiler accepts 
```

we'd be told that something is amiss. 

There are other confirming representation clauses as well. Thinking 
again of enumeration types, the underlying numeric values are 
integers, starting with zero and monotonically increasing from there 
up to N-1, where N is the total number of enumeral values. 

For example: 

.. code-block:: ada

   type Commands is (Off, On);
   
   for Commands use (Off => 0, On => 1);

As a result, Off is encoded as 0 and On as 1. That specific underlying 
encoding is guaranteed by the language, as of Ada 95, so this is just a 
confirming representation clause nowadays. But it was not guaranteed in 
the original version of the language, so if you wanted to be sure of the 
encoding values you would have specified the above. It wasn't confirming 
before Ada 95, in other words. 

But let's also say that the underlying numeric values are not what you 
want because you're interacting with some device and the commands are 
encoded with values other than 0 and 1. Maybe you want to use an 
enumeration type because you want to specify all the possible values 
actually used by clients. If you just used some numeric type instead and 

> made up constants for On and Off, there's nothing to keep clients from 

Reviewer: ```suggestion
made up constants for :ada:`On` and :ada:`Off`, there's nothing to keep clients from 
```

   Silly : constant Commands := 3;  -- won't compile

The storage and performance issues bring us back to confirming clauses. 
We want the compiler to recognize them as such, so that it can generate 
the usual code, thereby avoiding the unnecessary portability and 
performance issues. Why would we have such a confirming clause now? It 
might be left over from the original version of the language, written 
before the Ada 95 change. Some projects have lifetimes of several 
decades, after all, and changing the code can be expensive (certified 
code, for example). Whether the compiler does recognize confimring 
clauses is a feature of the compiler implementation. We can expect a 
mature compiler to do so, but there's no guarantee. 

Now let's turn to what is arguably the most common representation 
specification, that of record type layouts. 

Recall from the discussion above that Ada compilers are allowed to 
reorder record components in physical memory. In other words, the 
textual order in the source code is not necessarily the physical order 
in memory. That's different from, say, C, where what you write is what 
you get, and you better know what you're doing. On some targets a 

> misaligned struct component access will perform very poorly, or even 

Reviewer: ```suggestion
misaligned :c:`struct` component access will perform very poorly, or even 
```

trap and halt, but that's not the C compiler's fault. There's no 
guarantee that the Ada compiler will reorder the components to avoid 
this problem either. Indeed, GNAT did not reorder components until 
relatively recently but does now, at least for the more egregious 
performance cases. It does this reordering silently, too, although there 
is a switch to have it warn you when it does. To prevent reordering, 

> GNAT defines a pragma named No_Component_Reorder that does what the name 

Reviewer: ```suggestion
GNAT defines a pragma named :ada:`No_Component_Reorder` that does what the name 
```

language-defined way to specify a record type's layout. 

The record layout specification consists of the storage places for some 
or all components, specified with a record representation clause. This 
clause specifies the order, position, and size of components (including 
discriminants, if any). 

The approach is to first define the record type, as usual, using any 
component order you like |mdash| you're about to specify the physical 
layout explicitly, in the next step. That said, defining the components 
in the same order as in the physical layout is convenient because you 
can copy-and-paste the components in that second step, as a start. 

Let's reuse that record type from the earlier discussion:

.. code-block:: ada

   type My_Int is range 1 .. 10;
   
   subtype S is Integer range 1 .. 10;
   
   type R is record
      M : My_Int;
      X : S;
      B : Boolean;
      C : Character;
   end record;
   
The resulting layout is like so, assuming the compiler doesn't reorder 
the components: 

.. image:: images/unoptimized-record-component-order.png
  :width: 600
  :alt: Memory allocated to a record with unoptimized layout
   

> As a result, R'Size will be 80 bits (10 bytes), but those last two bytes 
> will be will be allocated to objects, for an Object_Size of 96 bits (12 

Reviewer: ```suggestion
As a result, :ada:`R'Size` will be 80 bits (10 bytes), but those last two bytes 
will be will be allocated to objects, for an :ada:`Object_Size` of 96 bits (12 
```

type declaration. The components' names appear, as in a record type 
declaration. But now, we don't repeat the components' types, instead we 
give their relative positions within the record, in terms of a relative 
offset that starts at zero. We also specify the bits we want them to 
occupy within the storage elements starting at that offset. 

.. code-block:: ada

   for R use record
      X at 0 range 0 .. 31;  -- note the order swap,
      M at 4 range 0 ..  7;  -- with this component
      B at 5 range 0 ..  7;
      C at 6 range 0 ..  7;
   end record;

Now we'll get the optimized order, and we'll always get that order, or 
the layout specification won't compile in the first place. In the 
following diagram, both layouts, the default, and the one resulting from 
the record representation clause, are depicted for comparison: 

.. image:: images/record_layout_with_rep_clause.png
  :width: 600
  :alt: Memory allocated to a record with optimized layout
  
R'Size will be 56 bits (7 bytes), but that last padding byte will also 
be allocated to objects, so the Object_Size will be 64 bits (8 bytes). 
  
Notice how we gave each component an offset, after the reserved word 

> "at". These offsets are in terms of storage elements, and specify their 

Reviewer: ```suggestion
:ada:`at`. These offsets are in terms of storage elements, and specify their 
```

positions within the record object as a whole. They are relative to the 
beginning of the memory allocated to the record object so they are 

> numbered starting at zero. We want the X component to be the very first 
> component in the allocated memory so the offset for that one is zero. 
> The M component, in comparison, starts at an offset of 4 because we are 
> allocating 4 bytes to the prior component X: bytes 0 through 3 
> specifically. M just occupies one storage element so the next component, 
> B, starts at offset 5. Likewise, component C starts at offset 6. 

Reviewer: ```suggestion
numbered starting at zero. We want the :ada:`X` component to be the very first 
component in the allocated memory so the offset for that one is zero. 
The :ada:`M` component, in comparison, starts at an offset of 4 because we are 
allocating 4 bytes to the prior component :ada:`X`: bytes 0 through 3 
specifically. :ada:`M` just occupies one storage element so the next component, 
:ada:`B`, starts at offset 5. Likewise, component :ada:`C` starts at offset 6. 
```

An individual component may occupy part of a single storage element, all 
of a single storage element, multiple contiguous storage elements, or a 
combination of those (i.e., some number of whole storage elements but 
also part of another). The bit "range" specifies this bit-specific 
layout, per component, by specifying the first and last bits occupied. 

> The X component occupies 4 complete 8-bit storage elements, so the bit 

Reviewer: ```suggestion
The :ada:`X` component occupies 4 complete 8-bit storage elements, so the bit 
```

such implicit components without giving you a way to refer to them. 
Otherwise they might go exactly where you want some other component to 
be placed, or overlap that place. 

The positions (offsets) and the bit numbers must be static, informally 
meaning that they are known at compile-time. They don't have to be 
numeric literals, though. Numeric constants would work, but literals are 
the most common by far. 

Note that the language does not limit support for component clauses to 
specific component types. They need not be one of the integer types, in 
particular. For example, a position can be given for components that are 
themselves record types, or array types. Even task types are allowed as 
far as the language goes, although the implementation might require a 
specific representation, such as the component taking no bits whatsoever 
(:ada:`0 .. -1`). There are restrictions that keep things sane, for 
example rules about how a component name can be used within the overall 
record layout construct, but not restrictions on the types allowed for 
individual components. For example, here is a record layout containing a 

> String component, arbitrarily set to contain 11 characters: 

Reviewer: ```suggestion
:ada:`String` component, arbitrarily set to contain 11 characters: 
```

.. code-block:: ada

   type R is record
      S : String (1 .. 11);
      B : Boolean;
   end record;

   for R use record
      S at 0  range 0 .. 87;
      B at 11 range 0 .. 7;
   end record;
   

> Component S is to be the first component in memory in this example, 
> hence the position offset is 0, for the first byte of S. Next, S is 11 
> characters long, or 88 bits, so the bit range is 0 .. 87. That's 11 
> bytes of course, so S occupies storage elements 0 .. 10. Therefore, the 
> next component position must be at least 11, unless there is to be a 
> gap, in which case it would be greater than 11. We'll place B 
> immediately after the last character of S, so B is at storage element 
> offset 11 and occupying all that one byte's bits. 

Reviewer: ```suggestion
Component :ada:`S` is to be the first component in memory in this example, 
hence the position offset is 0, for the first byte of :ada:`S`. Next, :ada:`S` is 11 
characters long, or 88 bits, so the bit range is 0 .. 87. That's 11 
bytes of course, so :ada:`S` occupies storage elements 0 .. 10. Therefore, the 
next component position must be at least 11, unless there is to be a 
gap, in which case it would be greater than 11. We'll place :ada:`B` 
immediately after the last character of :ada:`S`, so :ada:`B` is at storage element 
offset 11 and occupying all that one byte's bits. 
```

      X : S;
      B : Boolean;
      C : Character;
   end record with
      Alignment => 1;

Alignment values are in terms of storage elements. The effect of the 
aspect or attribute clause is to ensure that the starting address of the 
memory allocated to objects of the type will be a multiple of the 
specified value. 

In fact, whenever we specify a record type layout we really should also 
specify the record type's alignment, even though doing so is optional. 
Why? The alignment makes a difference in the overall record object's 
size. We've seen that already, with the padding bytes: the compiler will 
respect the alignment requirements of the components, and may add 
padding bytes within the record and also at the end to ensure components 
start at addresses compatible with their alignment requirements. The 
alignment also affects the size allocated to the record type even when 
the components are already aligned. As a result the overall size could 
be larger than we want for the sake of space. Additionally, when we pass 
such objects to code written in other languages, we want to ensure that 
the starting address of these objects is aligned as the external code 

> expects. The compiler might not chose that required alignment by 

Reviewer: ```suggestion
expects. The compiler might not choose that required alignment by 
```

default. 

Specifying alignment for record types is so useful that in the first 
version of Ada there was no syntax to specify alignment for anything 

> other than record types (via the obsolete "at mod" clause on record 

Reviewer: ```suggestion
other than record types (via the obsolete :ada:`at mod` clause on record 
```

representation clauses). 

> For that reason GNAT provides a pragma named Optimize_Alignment. This is 

Reviewer: ```suggestion
For that reason GNAT provides a pragma named :ada:`Optimize_Alignment`. This is 
```

a configuration pragma that affects the compiler's choice of default 
alignments where no alignment is explicitly specified. There is a 
time/space trade-off in the selection of these values, as we've seen. 
The normal choice tries to balance these two characteristics, but with 
an argument to the pragma you can give more weight to one or the other. 
The best approach is to specify the alignments explicitly, per type, for 
those that require specific alignment values. The pragma has the nice 
property of giving general guidance to the compiler for what should be 
done for the other types and objects not explicitly specified. 

Now let's look into the details. We'll use a case study for this 
purpose, including specifying sizes as well as alignments. 

> The code for the case study is as follows. It uses Size clauses to 
> specify the Sizes, instead of the Size aspect, just to emphasize that 

Reviewer: ```suggestion
The code for the case study is as follows. It uses :ada:`Size` clauses to 
specify the :ada:`Size`s, instead of the :ada:`Size` aspect, just to emphasize that 
```

> the the clause approach is not obsolete. 

Reviewer: ```suggestion
the :ada:`Size` clause approach is not obsolete. 
```

.. code-block:: ada

   package Some_Types is
   
      type Temperature is range -275 .. 1_000;
   
      type Identity is range 1 .. 127;
   
      type Info is record
         T  : Temperature;
         Id : Identity;
      end record;
   
      for Info use record
         T  at 0 range 0 .. 15;
         Id at 2 range 0 .. 7;
      end record;
   
      for Info'Size use 24;
   
      type List is array (1 .. 3) of Info;
      for List'Size use 24 * 3;
   
   end Some_Types;
     
When we compile this, the compiler will complain that the size for 

> "List" is too small, i.e., that the minimum allowed is 96 bits instead 

Reviewer: ```suggestion
:ada:`List` is too small, i.e., that the minimum allowed is 96 bits instead 
```

of the 72 we specified. We specified 24 * 3 because we said the record 
size should be 24 bits, and we want our array to contain 3 record 
components of that size, so 72 seems right.

What's wrong? As we've shown earlier, specifying the record type size 
doesn't necessarily mean that objects (in this case array components) 
are that size. The object size could be bigger than we specified for the 
type. In this case, the compiler says we need 96 total bits for the 
array type, meaning that each of the 3 array components is 32 bits wide 
instead of 24. 

Why is it 32 bits? Because the alignment for Info is 2 (on this 
machine). The record alignment is a multiple of the largest 
alignment of the enclosed components. The alignment for type Temperature 
(2), is larger than the alignment for type Identity (1), therefore the 
alignment for the whole record type is 2. We need to go from that number 
of storage elements to a number of bits for the size. 

Here's where it gets subtle. The alignment is in terms of storage 
elements. Each storage element is of a size in bits given by 

> System.Storage_Unit. We've said that on our hypothetical machine 
> Storage_Unit is 8, so storage elements are 8 bits wide on this machine. 

Reviewer: ```suggestion
:ada:`System.Storage_Unit`. We've said that on our hypothetical machine 
:ada:`Storage_Unit` is 8, so storage elements are 8 bits wide on this machine. 
```

Bytes, in other words. Therefore, to get the required size in bits, we 
have to find a multiple of the two 8-bit bytes (specified by the 

> alignment) that has at least the number of bits we gave in the Size 

Reviewer: ```suggestion
alignment) that has at least the number of bits we gave in the :ada:`Size` 
```

clause. Two bytes only provides 16 bits, so that's not big enough, we 
need at least 24 bits. The next multiple of 2 bytes is 4 bytes, 
providing 32 bits, which is indeed larger than 24. Therefore, the 
overall size of the record type, consistent with the alignment, is 4 
bytes, or 32 bits. That's why the compiler says each array component is 
32 bits wide. 

But for our example let's say that we really want to use only 72 total 
bits for the array type (and that we want three array components). 
That's the size we specified, after all. So how do we get the record 
type to be 24 bits instead of 32? Yes, you guessed it, we change the 
alignment for the record type. If we change it from 2 to 1, the size of 
24 bits will work. Adding this Alignment clause line will do that: 

.. code-block:: ada

   for Info'Alignment use 1;

An alignment of 1 means that any address will work, assuming that 
addresses refer to entire storage elements. (An alignment of 0 would 
mean that the address need not start on a storage element boundary, but 
we know of no such machines.) 

> We can even entirely replace the Size clause with the Alignment clause, 
> because the Size clause specifying 24 bits is just confirming: it's the 
> value that 'Size would return anyway. The problem is the object size. 

Reviewer: ```suggestion
We can even entirely replace the :ada:`Size` clause with the :ada:`Alignment` clause, 
because the :ada:`Size` clause specifying 24 bits is just confirming: it's the 
value that :ada:`'Size` would return anyway. The problem is the object size. 
```

Now, you may be wondering why an alignment of 1 would work, given that 

> the alignment of the Temperature component is 2. Wouldn't it slow down 

Reviewer: ```suggestion
the alignment of the :ada:`Temperature` component is 2. Wouldn't it slow down 
```

the code, or even trap? Well, maybe. It depends on the machine. If it 
doesn't work we would just have to use 32 bits for the record type, with 
the original alignment of 2, for a larger total array size. 

We said earlier that there are only a small number of reasons to specify 
:ada:`'Size` for a type. We can mention one of them now. Setting 'Size 
can be useful to give the minimum number of bits to use for a component 
of a packed composite type, that is, within either a record type or an 

> array type that is explicitly packed via the aspect or pragma Pack. It 

Reviewer: ```suggestion
array type that is explicitly packed via the aspect or pragma :ada:`Pack`. It 
```

No, it isn't earth-shattering, but other uses are more valuable, to be 
discussed soon. 

One thing we will leave unaddressed (pun intended) is the question of 
bit ordering and byte ordering within our record layouts. In other 
words, the "endian-ness". That's a subject beyond the scope of this 
course. Suffice it to say that GNAT provides a way to specify record 
layouts that are independent of the endian-ness of the machine, within 
some implementation-oriented limits. That's obviously useful when the 
code might be compiled for a different ISA in the future. On the other 
hand, if your code is specifically for a single ISA, e.g. Arm, even if 
different boards and hardware vendors are involved, there's no need to 
be independent of the endian-ness. It will always be the same in that 
case. (Those are "famous last words" though.) 

Although specifying record type layouts and alignments are perhaps the 
most common representation characteristics expressed, there are a couple 
of other useful cases. Both involve storage allocation.

One useful scenario concerns tasking. We can specify the number of 
storage elements reserved for the execution of a task object, or all 

> objects of a task type. You use the Storage_Size aspect to do so: 

Reviewer: ```suggestion
objects of a task type. You use the :ada:`Storage_Size` aspect to do so: 
```

.. code-block:: ada

   task Servo with
     Storage_Size => 1 * 1024,
     ...
     
Or the corresponding pragma:

.. code-block:: ada

   task Servo is
      pragma Storage_Size (1 * 1024);
   end Servo;     

The aspect seems textually cleaner and lighter unless you have task 
entries to declare as well. In that case the line for the pragma 
wouldn't add all that much. That's a matter of personal aesthetics 
anyway. 

The specified number of storage elements includes the size of 
the task's stack (GNAT does have one, per task). The language does not 
specify whether or not it includes other storage associated with the 
task used for implementing and managing the task execution. With GNAT, 
the extent of the primary stack size is the value returned, ignoring any 
other storage used internally in the run-time library for managing the 
task. 

The GNAT run-time library allocates a default stack amount to each task, 
with different defaults depending on the underlying O.S., or lack 
thereof, and the target. You need to read the documentation to find the 
actual amount, or, with GNAT, read the code (or just ask for help via 

> your GNAT account number). 

Reviewer: I believe we shouldn't mention the "account number" in this book.

Author: Thanks!

You would need to specify this amount in order to either increase or 
decrease the allocated storage. If the task won't run properly, perhaps 
crashing at strange and seemingly random places, there's a decent chance 
it is running out of stack space. That might also be the reason if you 
have a really deep series of subprogram calls that fails. The correction 
is to increase the allocation, as shown above. How much? Depends on the 
application code. Try increasing it until it runs properly. Then, 
iteratively decrease it a little at a time until it starts to fail 
again. Add a little back until it runs, and leave it there. 

Even if the task doesn't seem to run out of task stack, you might want 
to reduce it anyway, to the extent possible, because the total amount of 
storage on your target might be limited. Some of the GNAT bare-metal 
embedded targets have very small amounts of memory available, so much so that 
the default task stack allocations would exhaust the memory available 
quickly. That's what the example above does: empirical data showed that 

> the Servo task could run with just 1K bytes allocated, so we reduced it 

Reviewer: ```suggestion
the :ada:`Servo` task could run with just 1K bytes allocated, so we reduced it 
```

from the default accordingly. (We specified the size with that 
expression for the sake of readability, relative to using literals 
directly.) 

Notice we said "empirical data" above. How do we know that we exercised 
the task's thread of control exhaustively, such that the arrived-at 
allocation value covers the worst case? We don't, not with certainty. If 
we really must know the allocation will suffice for all cases, say 

> because this is a high-integrity application, we would use GNATstack. 

Reviewer: ```suggestion
because this is a high-integrity application, we would use :program:`GNATstack`. 
```

known at compile time. It does not rely on empirical run-time 
information.

The other useful scenario for allocating storage concerns access types, 
specifically access types whose values designate objects, as opposed to 
designating subprograms. (Remember, objects are either variables or 
constants.) There is no notion of dynamically allocating procedures and 
functions in Ada so access-to-subprogram types are not relevant here. 
But objects can be of protected types (or task types), and protected 
objects can "contain" entries and protected subprograms, so there's a lot 
of expressive power available. You just don't dynamically 
allocate procedures or functions as such. 

First, a little background.

By default, the implementation chooses a standard storage pool for each 
access-to-object type. The storage allocated by an allocator (i.e., 
:ada:`new`) for a given access-to-object type comes from the associated 
pool. 

Several access types can share the same pool. By default, the 
implementation might choose to have a single global storage pool, used 
by all such access types. This global pool might consist merely of calls 

> to operating system routines (e.g., "malloc"), or it might be a 

Reviewer: ```suggestion
to operating system routines (e.g., :c:`malloc`), or it might be a 
```

choose to create a new pool for each access-to-object type, reclaiming 
the pool's memory when the access type goes out of scope (if ever). 
Other schemes are possible. 

Finally, users may define new pool types, and may override the choice of 
pool for an access-to-object type by specifying :ada:`Storage_Pool` for 
the type. In this case, allocation (via :ada:`new`) takes memory from 
the user-defined pool and deallocation puts it back into that pool, 
transparently. 

With that said, here's how to specify the storage to be used for an 
access-to-object type. There are two ways to do it.

If you specify :ada:`Storage_Pool` for an access type, you indicate a 
specific pool object to be used (user-defined or vendor-defined). The 
pool object determines how much storage is available for allocation via 
:ada:`new` for that access type. 

Alternatively, you can specify :ada:`Storage_Size` for the access type. 
In this case, an implementation-defined pool is used for the access 
type, and the storage available is at least the amount requested, maybe 
more (it might round up to some advantageous block size, for example). 

> If the implementation cannot satisfy the request, Storage_Error is 

Reviewer: ```suggestion
If the implementation cannot satisfy the request, :ada:`Storage_Error` is 
```
